docs/language-specification.md

```diff
@@ -0,0 +1,204 @@
+# p4-constraints language specification
+
+The general design philosophy behind the p4-constraints language is to recycle
+[P4 language] constructs as much as possible, both syntactically and
+semantically. Where the languages overlap, p4-constraints aims to be consistent
+with the [P4 language specification]. This is to minimize the cost of adaption
+for programmers already familiar with [P4].
+
+Where [P4] lacks constructs required in p4-constraints, we add new constructs
+guided by the principle of least surprise.
+
+[P4]: https://p4.org/specs/
+[P4 language]: https://p4.org/specs/
+[P4 language specification]: https://p4.org/specs/
+
+## Tour of the language
+
+The p4-constraints language is a simple, expression-based language for
+specifying properties of P4 table entries.
+
+### Keys and comparisons
+
+At the heart of the language lies the ability to talk about table keys,
+and to constrain their values using comparison operators.
+```p4
+@entry_restriction("ipv4.valid == 1")
+table ipv4_table {
+  key = {
+    ipv4.isValid() : exact @name("ipv4.valid");
+    ipv4.dst : ternary;
+  }
+}
+```
+For example, the above constraint states that the key `ipv4.valid` must be equal
+to `1` for all table entries. Just like P4, the constraint language provides the
+comparison operators `==`, `!=`, `>`, `>=`, `<`, and `<=` with the usual
+semantics.
+
+### Accessing the fields of a key
+
+While an `exact` match consists of a single value, a `ternary` match is given
+by a value together with a mask. Similarly, an `lpm` match consists of a value
+together with a prefix length, and a range match consists of lower and upper
+bounds.
+
+To access the different values associated with a key, the language provides a
+field access (or "projection") operator `::`.
+```p4
+@entry_restriction("ipv6.dst::prefix_length <= 64")
+table ipv6_table {
+  key = {
+    ipv6.dst : lpm;
+  }
+}
+```
+For example, the above constraint states that entries should only ever match on
+the most-significant 64 bits of an IPv6 address.
+
+More generally, assume we have a P4 table of the form
+```p4
+table t {
+  key = {
+    k : <match_type>;
+  }
+}
+```
+with `k` being of type `bit<W>`. Then the fields that can be accessed using
+the `::` operator together with their types are summarized in the following
+table:
+
+| <match_type> | field            | field type |
+|--------------|------------------|------------|
+| exact        | k::value         | bit\<W\>   |
```

Comment:
Do we support optional?

Reply:
No, it hadn't made its way into p4runtime yet at the time.
I opened [an issue](https://github.com/p4lang/p4-constraints/issues/20).